Emoncms history component

joyrider3774 opened the pull request:
**Description:**

Emoncms history compnent is a history component just like / based on the dweet.io history component and has the same functionality but sends the data to emoncms.

This would for example let people push there power readings from zwave power devices to emoncms for further processing. 

But can be used to send any (numerical) data

**Related issue (if applicable):** fixes #

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#

https://github.com/home-assistant/home-assistant.github.io/pull/991

**Example entry for `configuration.yaml` (if applicable):**

``` yaml
# Example configuration.yaml entry
emoncms_history:
  api_key: put your emoncms WRITE api key here
  url: https://emoncms.org
  inputnode: 19
  whitelist:
    - sensor.owm_temperature
    - sensor.owm_wind_speed
    - sensor.owm_humidity
    - sensor.mold_indicator
    - sensor.emoncms1_feedid_29
    - sensor.cpu_use
    - binary_sensor.pir_motion
```

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
- [X] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If code communicates with devices, web services, or a:
- [N/A] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  tox doesn't run on windows
- [N/A] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  no new dependcies
- [N/A] New dependencies are only imported inside functions that use them ([example][ex-import]).
   no new dependcies
- [N/A] New dependencies have been added to `requirements_all.txt` by running 
  `script/gen_requirements_all.py`.
   no new dependcies 
- [X] New files were added to `.coveragerc`.

If the code does not interact with devices:
- [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
- [ ] Tests have been added to verify that the new code works.

balloob commented on homeassistant/components/emoncms_history.py:
```diff
@@ -0,0 +1,92 @@
+        vol.Required(CONF_URL): cv.string,
+        vol.Required(CONF_INPUTNODE): cv.positive_int,
+        vol.Required(CONF_WHITELIST, default=[]):
+            vol.All(cv.ensure_list, [cv.entity_id]),
```
Probably want to add `vol.Length(min=1)` after `[cv.entity_id]`.

joyrider3774 replied:
i switched to vol.Required(CONF_WHITELIST): cv.entity_ids, which handles everything that's needed the intial changes i did had problem when the user would only write `whitelist:` and no other values giving errors about a nonetype. cv.entity_ids checks for this

balloob commented on homeassistant/components/emoncms_history.py:
```diff
@@ -0,0 +1,92 @@
+        vol.Required(CONF_API_KEY): cv.string,
+        vol.Required(CONF_URL): cv.string,
+        vol.Required(CONF_INPUTNODE): cv.positive_int,
+        vol.Required(CONF_WHITELIST, default=[]):
```
Don't add a default value unless you want people to be able to whitelist 0 entities. (judging on the code, you don't)

balloob commented on homeassistant/components/emoncms_history.py:
```diff
@@ -0,0 +1,92 @@
+        send_data(conf.get(CONF_URL), conf.get(CONF_API_KEY),
+                  str(conf.get(CONF_INPUTNODE)), json_str)
+
+    hass.bus.listen(EVENT_STATE_CHANGED, emoncms_event_listener)
```
Just do `track_state_change(hass, whitelist)`

joyrider3774 replied:
am doing this now

balloob commented on homeassistant/components/emoncms_history.py:
```diff
@@ -0,0 +1,92 @@
+            return
+
+        json_body[state.entity_id] = _state
+        json_str = json.dumps(json_body, separators=(",", ":"))
```
What kind of voodoo are these lines? How does the end result look and why?

joyrider3774 replied:
the input api of emoncms does not accept normal json or at least it would not work if itried it
They require data to be send like in this format for example
http://192.168.0.101:8083/emoncmsnew/input/post.json?node=22&json={input1:10,input2:10,input3:30}
so those lines are there to convert the data so it becomes a format like this : `{key1:value1,key2:value2,key3:value3}` 
it doesnt accept "key1":"value1" nor does it accept valueswith spaces after the `:` and `,`

balloob commented on homeassistant/components/emoncms_history.py:
```diff
@@ -0,0 +1,92 @@
+    return True
+
+
+@Throttle(MIN_TIME_BETWEEN_UPDATES)
```
You probably don't want this. This means that if 2 entities change state within 1 second, the second one will get discarded.

joyrider3774 replied:
i initially knew about this and had put a warning in the documentation about this, but i removed it now as well as in the documentation. i only added the 1 second throttle to not hammer the emoncms site (in case it was the cloud version) but it has a negative impact. For reference the dweet.io history component had this same limitation

tinglis1 replied:
An alternative, but not sure how to do it, would be to cache the data locally for a period and send the states with time stamps periodically. From memory the emoncms API supports this.

joyrider3774 replied:
@tinglis1 as far as i could see the bulk data does not support providing a key name which i do here. The key names i send are the entityid's so i just use one node unique to hass and all it's values are keys with the name and value for it.

if you look at [api](https://www.emoncms.org/site/api#input) the bulk data section you'll notice there's no mention of providing keynames when sending bulk data so you can't really identify the values afterwards which imo is needed & handy

tinglis1 replied:
@joyrider3774 Very true.

I had a look at the emoncms code and it should be straight forward enough (famous last words) to add an optional key name to the bulk data without breaking the existing API.
https://github.com/emoncms/emoncms/blob/master/Modules/input/input_controller.php

I am not overly familiar with php but this change might be within my ability. It won't help you here, but might be good for future as it has been requested a few times. 
https://openenergymonitor.org/emon/node/2249

This would be pretty low on my list of things to do so probably won't get to it for some time.

balloob commented on homeassistant/components/emoncms_history.py:
```diff
@@ -0,0 +1,92 @@
+        req = requests.get(fullurl, params={"apikey": apikey,
+                                            "node": node,
+                                            "json": jsonstr},
+                           verify=False, allow_redirects=True,
```
`verify=False` -> Nope ⚠️

joyrider3774 replied:
this is removed now

joyrider3774 commented:
crap did i mess something up ? with all those commits from other users in my branch ? i tried to do a rebase but i had never done that before and tortoise came saying about a merge conflict that i had to pull first before doing a push which i did but then i had to fix conflicts on my own file somehow ...

Should i create a new pr ?

joyrider3774 committed:
Emoncms_history component, fix git mess

joyrider3774 commented:
i managhed to fix it thanks to @w1ll1am23 but the history is gone. I changed everything you requested, except for the jsondump voodoo. You probably want to go over the file gain to see if it's all ok now. for fixing the git mess i lost the history of my commits

joyrider3774 commented on homeassistant/components/emoncms_history.py:
```diff
@@ -0,0 +1,88 @@
+    # as inactive and possibly send emails that feeds based
+    # on the input value have not changed (if the user enabled
+    # that with the extra, optional scripts)
+    json_body = {}
```
I might have to rewite the sensor then as i did not think of this and emoncms expects data to be received at regular intervals. I just tested by only sending data for one specific entityid when it's state changes and it does not work well with emoncms (graphs don't show etc). So Instead of using track_state_change i might have to switch to track_point_in_time and query all specified whitelisted entities for it's state at a regular interval (like 30 seconds) and send the data also at a regular interval. So then i still send all whitelisted entities at a regular interval but with their correct state at that moment

joyrider3774 committed:
- switch to track_point_in_time to send all data at foxed interval
- don't use json_dump
- switch to http post instead of http get

joyrider3774 commented:
i switched to track_point_in_time to send data at a fixed interval, emoncms expects this for some of their feed mechanisms (only one accepts non fixed interval). 
Another reason is that interpollation in graphs is not always wanted. Imagine the following scenario:
you have a sensor that has a state of 10 at 6 am and changes to 20 at 6pm if i would use track_state_change and only send the value when the states change the graphs would interpollate between the 2 points while it should not because at 5:59pm the value was still 10. 
i also changed to http post instead of http get to be sure i won't have any problems with url limiting factors when more sensors are whitelisted

balloob commented:
🐬
